### PR TITLE
Disallow parenthesized specializers in rule heads

### PR DESCRIPTION
**Pull request by gj**
Effectively reverts #61.

Makes all of these cases parse errors:

```py
polar.register_constant(1, name="Foo")
polar.load_str("f(_: (Foo));")
assert qeval("f(1)")
assert not qeval("f(2)")
```

```py
polar.load_str("""
    type f(_: (Foo));
    f(_: (Foo));
""")
```

```py
with pytest.raises(exceptions.ValidationError):
    polar.load_str("""
        type f(_: (Foo));
        f(_: Foo);
    """)
```

```py
polar.register_constant(1, name="Foo")
polar.register_constant(1, name="Bar")
with pytest.raises(exceptions.ValidationError):
    polar.load_str("""
        type f(_: (Bar));
        f(_: (Foo));
    """)
```

PR checklist:

- [x] Added changelog entry.

### CODE REVIEW

**Review by samscott89 — COMMENTED** [review]
Draft approved

File: docs/content/any/project/changelogs/NEXT.md [hunk 1]
```diff
@@ -12,7 +12,29 @@ draft: true
...
+  before upgrading.
+{{% /callout %}}
+
+##### Parenthesized variable dereferencing in specializer position is no longer permitted
+
+Previously, Oso would treat parenthesized specializers as variables. This made
+it possible to write rules with parameters that referenced each other, like so:
+
+```polar
+isa(x, y, x: (y));
```
Comment (gw000) [inline docs/content/any/project/changelogs/NEXT.md:28]: Is the current behavior equivalent to `isa(x, y, x) if x matches y;`? If so we could call that out here so anyone who's affected by this change will know how to fix it.

Comment (gj) [reply docs/content/any/project/changelogs/NEXT.md:28]: I think it would be equivalent to `isa(x, y) if x matches y;` — the `isa/3` rule in the changelog is complemented by an `isa(x, y) if isa(x, y, x);` in @samscott89's original PR (#61)

Comment (gj) [reply docs/content/any/project/changelogs/NEXT.md:28]: de38bb57

Comment (samscott89) [reply docs/content/any/project/changelogs/NEXT.md:28]: So `isa` is actually a token in the lexer. So that example doesn't even work -- you get a parse error on unexpected token

Also, the rule doesn't even work any more. We'll try and unify x with y (since `y` isn't picked up as a tag, it's an external instance).

And the replacement suggestion doesn't work either. `isa(x, y) if x matches y` will actually do `x matches y{}` and you get `unregistered class y`.

And we also don't seem to support `x matches (y)` so that alternative doesn't work.

In conclusion... doing `rule(x, y, x: (y))` maybe _is_ pathological... it's a _really_ horrible way to write `rule(x, y, x)`. And there is nothing useful you can currently do with that syntax.

Comment (samscott89) [reply docs/content/any/project/changelogs/NEXT.md:28]: Concretely. I'd remove the existing changelog entry, and replace it with a super minor entry like: "removed syntax for `f(x: (y))` which does not currently achieve anything".

Comment (samscott89) [reply docs/content/any/project/changelogs/NEXT.md:28]: Oh, fun fact. The test cases with `x: (y)` in them output `x matches y` as constraints. But if you ground either side it'll ultimately end up attempting to do `x = y`.

**Review by samscott89 — CHANGES_REQUESTED** [review]
All looks fine, other than the changelog entry: https://github.com/osohq/oso/pull/1356/files#r743918189

File: docs/content/any/project/changelogs/NEXT.md (hunk 1 above)
Comment (gj) [reply docs/content/any/project/changelogs/NEXT.md]: > Concretely. I'd remove the existing changelog entry, and replace it with a super minor entry like: "removed syntax for `f(x: (y))` which does not currently achieve anything".

256831d5

**Review by samscott89 — APPROVED** [review]
